src/Features/Core/Portable/EmbeddedLanguages/StackFrame/StackFrameNodeDefinitions.cs

```diff
@@ -0,0 +1,426 @@
+        {
+            Debug.Assert(dotToken.Kind == StackFrameKind.DotToken);
+
+            Left = left;
+            DotToken = dotToken;
+            Right = right;
+        }
+
+        internal override int ChildCount => 3;
+
+        public override void Accept(IStackFrameNodeVisitor visitor)
+            => visitor.Visit(this);
+
+        internal override StackFrameNodeOrToken ChildAt(int index)
+            => index switch
+            {
+                0 => Left,
+                1 => DotToken,
+                2 => Right,
+                _ => throw new InvalidOperationException()
+            };
+    }
+
+    /// <summary>
+    /// The simplest identifier node, which wraps a <see cref="StackFrameKind.IdentifierToken" />
+    /// </summary>
+    internal sealed class StackFrameIdentifierNameNode : StackFrameSimpleNameNode
+    {
+        internal override int ChildCount => 1;
+
+        public StackFrameIdentifierNameNode(StackFrameToken identifier)
+            : base(identifier, StackFrameKind.TypeIdentifier)
+        {
+        }
+
+        public override void Accept(IStackFrameNodeVisitor visitor)
+            => visitor.Visit(this);
+
+        internal override StackFrameNodeOrToken ChildAt(int index)
+            => index switch
+            {
+                0 => Identifier,
+                _ => throw new InvalidOperationException()
+            };
+    }
+
+    /// <summary>
+    /// An identifier with an arity, such as "MyNamespace.MyClass`1" 
+    /// </summary>
+    internal sealed class StackFrameGenericNameNode : StackFrameSimpleNameNode
+    {
+        /// <summary>
+        /// The "`" token in arity identifiers. Must be <see cref="StackFrameKind.GraveAccentToken"/>
+        /// </summary>
+        public readonly StackFrameToken GraveAccentToken;
+
+        public readonly StackFrameToken NumberToken;
+
+        internal override int ChildCount => 3;
+
+        public StackFrameGenericNameNode(StackFrameToken identifier, StackFrameToken graveAccentToken, StackFrameToken numberToken)
+            : base(identifier, StackFrameKind.GenericTypeIdentifier)
+        {
+            Debug.Assert(graveAccentToken.Kind == StackFrameKind.GraveAccentToken);
+            Debug.Assert(numberToken.Kind == StackFrameKind.NumberToken);
+
+            GraveAccentToken = graveAccentToken;
+            NumberToken = numberToken;
+        }
+
+        public override void Accept(IStackFrameNodeVisitor visitor)
+            => visitor.Visit(this);
+
+        internal override StackFrameNodeOrToken ChildAt(int index)
+            => index switch
+            {
+                0 => Identifier,
+                1 => GraveAccentToken,
+                2 => NumberToken,
+                _ => throw new InvalidOperationException()
+            };
+    }
+
+    /// <summary>
+    /// Represents an array type declaration, such as string[,][]
+    /// </summary>
+    internal sealed class StackFrameArrayTypeExpression : StackFrameNameNode
```

Comment:
ah, found an issue.  this isn't really roslyn shape.  Do this:

```
abstract class StackFrameTypeNode : StackFrameNode { ... }
abstract class StackFrameNameNode : StackFrameTypeNode { ... }
class StackFrameArrayTypeNode : StackFrameTypeNode  { ... }
class StackFrameParameterDeclarationNode { StackFrameTypeNode Type; }
```

Reply:
> abstract class StackFrameNameNode : StackFrameTypeNode { ... }

This seems wrong to me. That would mean that `StackFrameIdentifierNameNode` is a `StackFrameTypeNode`

Reply:
oh actually this may be fine. I forgot that `StackFrameParameterDeclarationNode` doesn't use `StackFrameNameNode` for the identifier. It explicitly requires a token

Reply:
> This seems wrong to me. That would mean that StackFrameIdentifierNameNode is a StackFrameTypeNode

yup, that's totally fine (see roslyn for example).  Say i have this:  `public Foo Bar() { }`.  `Foo` is an identifier, but that's definitely a Type context.  In other words *all* identifiers/name can be used where a Type is needed (though not vice versa).